server/cluster/unsafe_recovery_controller.go

```diff
@@ -337,27 +359,40 @@ func (u *unsafeRecoveryController) HandleStoreHeartbeat(heartbeat *pdpb.StoreHea
 				}
 				fallthrough
 			case createEmptyRegion:
-				if u.generateCreateEmptyRegionPlan(newestRegionTree, peersMap) {
+				hasPlan, err = u.generateCreateEmptyRegionPlan(newestRegionTree, peersMap)
+				if err != nil {
+					break
+				}
+				if hasPlan {
 					u.changeStage(createEmptyRegion)
 					break
 				}
 				fallthrough
 			case exitForceLeader:
 				// no need to generate plan, empty recovery plan triggers exit force leader on TiKV side
-				if u.generateExitForceLeaderPlan() {
+				hasPlan = u.generateExitForceLeaderPlan()
+				if hasPlan {
 					u.changeStage(exitForceLeader)
 				}
 			default:
 				panic("unreachable")
 			}
 
-			hasPlan := len(u.storeRecoveryPlans) != 0
-			if u.err != nil {
-				if u.handleErr() {
+			if err != nil {
+				if u.HandleErr(err) {
 					return
 				}
+				u.storePlanExpires = make(map[uint64]time.Time)
+				u.storeRecoveryPlans = make(map[uint64]*pdpb.RecoveryPlan)
+				// Clear the reports etc.
+				u.changeStage(exitForceLeader)
```

Comment:
should we reset timeout as well?

Reply:
It makes sense to reset the timeout if the failure is caused by "timeout". For rest of the errors, we probably should leave the timeout as it is, do you agree?